### PR TITLE
C "memcpy" shim: ensure the pointers are valid

### PR DESCRIPTION
**Pull request by RalfJung**
Also add tests for some other shims that already behave correctly

Fixes https://github.com/rust-lang/miri/issues/2966

### CODE REVIEW

**Comment by oli-obk** [comment]
I don't get this CI failure. `memrchr` exists (it works locally, and it doesn't have any cargo features/cfgs/anything).

**Comment by oli-obk** [comment]
Ooooh, that wasn't clear from the ui_test annotation: this is for a cross compilation target.

**Comment by RalfJung** [comment]
Btw, what is this in the test output?
```
   error file=tests/fail/shims/memrchr_null.stderr,line=0,title=actual output differs from expected::```diff%0A1 unchanged lines skipped%0A-error: Undefined Behavior: memory access failed: null pointer is a dangling pointer (it has no provenance)%0A+error[E0425]: cannot find function `memrchr` in crate `libc`%0A   --> $DIR/memrchr_null.rs:LL:CC%0A    |%0A LL |         libc::memrchr(ptr::null(), 0, 0);%0A-   |         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ memory access failed: null pointer is a dangling pointer (it has no provenance)%0A+   |               ^^^^^^^%0A    |%0A-   = help: this indicates a bug in the program: it performed an invalid operation, and caused Undefined Behavior%0A-   = help: see https://doc.rust-lang.org/nightly/reference/behavior-considered-undefined.html for further information%0A-   = note: BACKTRACE:%0A-   = note: inside `main` at $DIR/memrchr_null.rs:LL:CC%0A+  ::: CARGO_REGISTRY/.../mod.rs:LL:CC%0A+   |%0A+LL |     pub fn memchr(cx: *const c_void, c: c_int, n: size_t) -> *mut c_void;%0A+   |     -------------------------------------------------------------------- similarly named function `memchr` defined here%0A+   |%0A+help: a function with a similar name exists%0A+   |%0A+LL |         libc::memchr(ptr::null(), 0, 0);%0A+   |               ~~~~~~%0A+help: consider importing this function%0A+   |%0A+LL + use core::slice::memchr::memrchr;%0A+   |%0A+help: if you import `memrchr`, refer to it directly%0A+   |%0A+LL -         libc::memrchr(ptr::null(), 0, 0);%0A+LL +         memrchr(ptr::null(), 0, 0);%0A+   |%0A %0A-note: some details are omitted, run with `MIRIFLAGS=-Zmiri-backtrace=full` for a verbose backtrace%0A-%0A error: aborting due to previous error%0A %0A+For more information about this error, try `rustc --explain E0425`.%0A %0A```
```
That's a very long line repeating the entire diff... some kind of debug output?

File: tests/fail/shims/memrchr_null.stderr [hunk 1]
```diff
@@ -1,15 +1,28 @@
-error: Undefined Behavior: memory access failed: null pointer is a dangling pointer (it has no provenance)
+error[E0425]: cannot find function `memrchr` in crate `libc`
```
Comment (RalfJung) [inline tests/fail/shims/memrchr_null.stderr:1]: Looks like the blessed output now says "cannot find function"? That doesn't seem right?^^

**Comment by oli-obk** [comment]
yea 🤦 I forgot to bless normally again

@bors r=RalfJung,oli-obk

**Comment by oli-obk** [comment]
> No need to have that wrong bless in the history at all...

the `gh pr checkout` command does not allow force pushing 🙃

**Pull request merged**